msprime/trees.py

```diff
@@ -159,6 +177,34 @@ def __str__(self):
 
 
 class EdgesetTable(_msprime.EdgesetTable):
+    """
+    Class for tables describing all edgesets in a tree sequence, of the form
+        left	right	parent	children
+        5.0	7.0	4	0,1
+        2.0	2.5	5	1,3
+        3.2	3.5	4	0,1
+        0.0	3.2	6	0,1,3
+    These describe the half-open genomic interval affected: `[left, right)`,
+    the `parent` and the `children` on that interval.
+
+    Requirements: to describe a valid tree sequence, a `EdgesetTable` (and
+    corresponding `NodeTable`, to provide birth times) must satisfy:
+        1. each list of children must be in sorted order,
+        2. any two edgesets that share a child must be nonoverlapping, and
+        3. the birth times of the `parent` in an edgeset must be strictly
+            greater than the birth times of the `children` in that edgeset.
+    Furthermore, for algorithmic requirements
+        4. the the table must be sorted by birth time of the `parent`, and
+        5. any two edgesets corresponding to the same `parent` must be nonoverlapping.
```

Comment:
Your example there is disallowed by (2.) No multiple inheritance.  The reason I'm including (5.) and calling it an algorithmic requirement is because 

        left    right   parent  children
        0.0     1.0     4       0,
        0.0     2.0     4       1,

is not contradictory but is disallowed.